### PR TITLE
polychromatic: init at 0.7.3

### PR DESCRIPTION
**Pull request by evanjs**
## Notes regarding branch selection
~- use dev-1.0.0 branch~
  ~- the 1.0.0 branch uses meson instead of a plain Makefile and I found it~ ~much easier to create an expression for this version of the repository~
  ~- according to [this comment](https://github.com/polychromatic/polychromatic/issues/230#issuecomment-585216003), it assumes a [PR](https://github.com/openrazer/openrazer/pull/888) from openrazer is merged.~

## Related:
~#82269 - Init OpenRGB~
#90691 - openrgb: init at 0.2

<del>

## Current issues (OUTDATED)

<summary>polychromatic-controller errors</summary>
<details>

```
Unable to get device states for Razer Huntsman Elite
Traceback (most recent call last):
  File ’/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/lib/python3.7/site-packages/polychromatic/backends/openrazer.py’, line 283, in get_device
    effect = str(zone_to_device[zone].effect)
AttributeError: ’RazerFX’ object has no attribute ’effect’

This probably indicates a bug, wrong OpenRazer version or improperly specified Chroma device:
Razer Huntsman Elite (0226:1532)
```

</details>

<summary>polychromatic-tray-applet errors</summary>
<details>

```
/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped:183: DeprecationWarning: Gtk.StatusIcon.set_name is deprecated
  self.indicator.set_name("polychromatic-tray-applet")
/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped:184: DeprecationWarning: Gtk.StatusIcon.set_from_file is deprecated
  self.indicator.set_from_file(self._get_tray_icon())
/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped:185: DeprecationWarning: Gtk.StatusIcon.set_visible is deprecated
  self.indicator.set_visible(True)
Unable to get device states for Razer Huntsman Elite
Traceback (most recent call last):
  File ’/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/lib/python3.7/site-packages/polychromatic/backends/openrazer.py’, line 283, in get_device
    effect = str(zone_to_device[zone].effect)
AttributeError: ’RazerFX’ object has no attribute ’effect’

This probably indicates a bug, wrong OpenRazer version or improperly specified Chroma device:
Razer Huntsman Elite (0226:1532)
/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped:129: DeprecationWarning: Gtk.ImageMenuItem.set_image is deprecated
  item.set_image(img)            # GTK3: Deprecated, no replacement!
Traceback (most recent call last):
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped", line 862, in <module>
    indicator.setup()
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped", line 194, in setup
    self.build_indicator()
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped", line 274, in build_indicator
    root.append(self.build_device_submenu(device, details)[0])
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped", line 357, in build_device_submenu
    submenu[1].append(self._create_effect_submenu(details, zone)[0])
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-tray-applet-wrapped-wrapped", line 486, in _create_effect_submenu
    if "random" in starlight_options:
NameError: name 'starlight_options' is not defined
```

</details>

I tried using the master branch for openrazer, but encountered the same issues.
Interestingly, I found mention in the source for features that require OpenRazer 2.8.0+ ([here](https://github.com/polychromatic/polychromatic/blob/5ca66b65c6d739e20c973b76a74054476e9c8cce/pylib/backends/openrazer.py#L279)).

I am not sure if this is a typo or referring to the current master branch.

Additionally, this is what happens if no `~/.config/polychromatic/custom_icons` directory is found when `polychromatic-controller` is launched
<summary><h4>Missing directory error<h4></summary>
<details>

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/nix/store/xb1mp0f16lv23rwhfdc1665v5r20jfmm-python3-3.7.6/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/nix/store/xb1mp0f16lv23rwhfdc1665v5r20jfmm-python3-3.7.6/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/bin/..polychromatic-controller-wrapped-wrapped", line 342, in start_app
    self.controller.initalise_app(self.version, self.versions, args.tab)
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/lib/python3.7/site-packages/polychromatic/controller.py", line 110, in initalise_app
    self.send_view_variable("CUSTOM_ICONS", pref.get_custom_icons())
  File "/nix/store/9hsbcqi59nqqwgv0v4nvxgmkpabi97hn-polychromatic-unstable-2020-03-10/lib/python3.7/site-packages/polychromatic/preferences.py", line 371, in get_custom_icons
    return os.listdir(path.custom_icons)
FileNotFoundError: [Errno 2] No such file or directory: '/home/evanjs/.config/polychromatic/custom_icons'
```

</details></del>

___

<del>

## Current Functionality
As it stands, brightness for connected devices can be changed, but you cannot switch between tabs, and much of the application is inaccessible.

</del>

###### Motivation for this change
Polychromatic provides a graphical interface to manage Razer devices using Openrazer

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @adisbladis @worldofpeace

### CODE REVIEW

File: pkgs/applications/misc/polychromatic/default.nix [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    };
+
+    pathsToLink = [ "/bin" "/etc" "/lib" "/share" ];
+
+    buildInputs = [
+      cairo
+      hicolor-icon-theme
+    ];
+
+    propagatedBuildInputs = [
```
Comment (adisbladis) [inline pkgs/applications/misc/polychromatic/default.nix]: `polychromatic-tray-applet` needs to be wrapped with `lsusub` from `usbutils`.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -17940,6 +17940,10 @@ in
 
   poly = callPackage ../data/fonts/poly { };
 
+  polychromatic = callPackage ../applications/misc/polychromatic {
```
Comment (adisbladis) [inline pkgs/top-level/all-packages.nix]: Please use `python3.pkgs.callPackage` to get dependencies from the python3 set.

**Review by adisbladis — CHANGES_REQUESTED** [review]

**Comment by evanjs** [comment]
Ah, looks like we're blocked on https://github.com/openrazer/openrazer/pull/888 being merged, as mentioned [here](https://github.com/polychromatic/polychromatic/issues/230#issuecomment-585216003).

**Comment by symphorien** [comment]
The derivation lack meta.

**Comment by cawilliamson** [comment]
Would love to see this merged ASAP - really missing the functionality.

**Comment by evanjs** [comment]
For @cawilliamson and others that need something in the meantime, please check out Razer Genie (#84812), which was recently merged.

**Comment by evanjs** [comment]
The development status of Polychromatic 1.0.0 is provided via the [roadmap](https://polychromatic.app/docs/roadmap/) if anyone is interested.

I will be keeping an eye on this and investigate the amount of effort required to get 1.0.0 packaged once it is made available in a non-alpha state.

**Comment by lah7** [comment]
Developer of the Polychromatic app here - just thought I'd let you know:

* The new codebase is now in `master`.
* It is no longer blocked by  openrazer/openrazer#888 (persistence feature was merged in OpenRazer `master` now, but this app no longer _needs_ it anyway)
* The application is now a PyQt5 app, so the [dependencies have changed](https://polychromatic.app/docs/dependencies/#v100).
* Some metadata (package description) [has changed](https://github.com/polychromatic/polychromatic/blob/master/debian/control#L26-L28) as the project's scope is evolving.

At time of writing:

* Consider this codebase alpha quality, with a version string of v0.6.0, but may work better then v0.3.12 for certain types of hardware.
* Bugs expected, since extensive testing on many distros/environments hasn't started.
* Has one feature regression from v0.3.12 (custom effect editor), but this should be restored _soon_.

**Comment by lah7** [comment]
Just a note - this software released a [new stable v0.7.0](https://github.com/polychromatic/polychromatic/releases/tag/v0.7.0) version. The dependencies might want looking over with its [docs](https://docs.polychromatic.app/dependencies/).

**Comment by evanjs** [comment]
This PR now targets v0.7.0, and builds without errors, but e.g. the controller fails to run.

I am not sure which `callPackage` variant I should be using/package set I should be importing (qt5, pythonPackages, etc.).

I am also unsure how the application should be wrapped, as the controller depends on QT5, while the tray applet requires `libappindicator-gtk3`, and so on.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by evanjs** [comment]
I have updated the PR with a bump to `0.7.2`, and the application now launches, but I am unsure of some of the functionality at the moment.

There is also the `Restart OpenRazer` functionality that we might want to patch.
Currently, it seems to require `psutil` (Python module) and `killall`

**Comment by Icy-Thought** [comment]
v0.7.3 was released 19 days ago. Might as well push forward that update rather than the current outdated version (v0.7.2) :smile:

**Comment by Icy-Thought** [comment]
Last commit should produce a working package. Tested it locally and Polychromatic seems to be working fine!

**Review by Icy-Thought — APPROVED** [review]

**Comment by jrgiacone** [comment]
How can I make and run this locally while waiting?

**Comment by jrgiacone** [comment]
The only Issue I get building is getting the tray working

**Comment by Icy-Thought** [comment]
I wonder if you are the only one facing this issue, @jrgiacone. I just ran `polychromatic-tray-applet` to confirm this and the tray executed just fine. No failure had been encountered during this rather short testing period, as you had experienced.

Proof:
image

Logs (`-v`):
```
[openrazer] Initialising Device Manager...
Initialising AppIndicator3...
Initialised AppIndicator3
Loading device data...
[openrazer] Persistence read unnecessary for Razer Basilisk X HyperSpeed (Zone: main)
[openrazer] Persistence read unnecessary for Razer Basilisk X HyperSpeed (Zone: main)
Creating menus...
/nix/store/x2kw4c0kls347j5hxbjpghkrblwl0hqr-polychromatic/bin/..polychromatic-tray-applet-wrapped-wrapped:225: DeprecationWarning: Gtk.ImageMenuItem.set_image is deprecated
  item.set_image(img)            # GTK3: Deprecated, no replacement!
- Razer Basilisk X HyperSpeed
Finished setting up applet.
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -24528,6 +24528,10 @@ with pkgs;
 
   poly = callPackage ../data/fonts/poly { };
 
+  polychromatic = libsForQt5.callPackage ../applications/misc/polychromatic {
+    inherit (pkgs) meson wrapGAppsHook;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:24532]:
```suggestion
```
That does not work? Thats unexpected.

File: pkgs/applications/misc/polychromatic/default.nix [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  nativeBuildInputs = with python3Packages; [
+    pyqt5
+    desktop-file-utils
+    qt5.wrapQtAppsHook
+    wrapGAppsHook
+    ninja
+    meson
+    sassc
+  ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/polychromatic/default.nix:89]: Normally we put nativeBuildInputs first 🙃

File: pkgs/applications/misc/polychromatic/default.nix [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  postPatch = ''
+    patchShebangs scripts
+
+    substituteInPlace scripts/build-styles.sh \
+      --replace '$(which sassc 2>/dev/null)' '${sassc}/bin/sassc' \
+      --replace '$(which sass 2>/dev/null)' '${sassc}/bin/sass'
+
+    substituteInPlace pylib/common.py --replace "/usr/share/polychromatic" "$out/share/polychromatic"
+
+    scripts/build-styles.sh
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/applications/misc/polychromatic/default.nix:46]: I meant only ``scripts/build-styles.sh``. The other patching lines should have stayed in postPatch. Sorry about being to vague.

**Comment by jrgiacone** [comment]
> 

@Icy-Thought do you mind seeing if I have a package missing, here is my config:
```
# Edit this configuration file to define what should be installed on
# your system.  Help is available in the configuration.nix(5) man page
# and in the NixOS manual (accessible by running ‘nixos-help’).

{ config, pkgs, ... }:

{
  imports =
    [ # Include the results of the hardware scan.
      ./hardware-configuration.nix
    ];

  # Brtfs subvolume options:
  fileSystems = {
    "/home".options = ["compress-force=zstd,ssd"];
    "/nix".options = ["compress-force=zstd,ssd,noatime" ];
    "/var/log".options = ["compress-force=zstd,ssd,noatime" ];
    "/etc".options = ["compress-force=zstd,ssd,noatime" ];
    "/root".options = ["compress-force=zstd,ssd,noatime"];
  };

  #boot.extraModprobeConfig = ''hid-playstation'';

  # Brtfs nixos options
  services.btrfs = {
    autoScrub = {
      enable = true;
      interval = "weekly";
      fileSystems = [ "/" ];
    };
  };

  # Options for the bootloader, currently using grub with menuentries for arch and windows:

  # Use the systemd-boot EFI boot loader.
  # boot.loader.systemd-boot.enable = true;
  # boot.loader.efi.canTouchEfiVariables = true;

  boot.loader = {
    #systemd-boot.enable = false;
    grub = {
      enable = true;
      copyKernels = true;
      efiSupport = true;
      devices = ["nodev"];
      version = 2;
      extraEntries = ''
        menuentry 'Arch Linux (on /dev/nvme1n1p3)' --class gnu-linux --class gnu --class os $menuentry_id_option 'osprober-gnulinux-/boot/vmlinuz-linux--ab300c2c-d7f8-4395-a8aa-27f8215394ae' {
                insmod part_gpt
                insmod ext2
                search --no-floppy --fs-uuid --set=root ab300c2c-d7f8-4395-a8aa-27f8215394ae
                linux /boot/vmlinuz-linux root=UUID=ab300c2c-d7f8-4395-a8aa-27f8215394ae rw loglevel=3 quiet rd.driver.blacklist=nouveau nvidia-drm.modeset=1
                initrd /boot/amd-ucode.img /boot/initramfs-linux.img
        }
        menuentry 'Windows Boot Manager (on /dev/nvme0n1p1)' --class windows --class os $menuentry_id_option 'osprober-efi-4E34-3A1B' {
                insmod part_gpt
                insmod fat
                search --no-floppy --fs-uuid --set=root 4E34-3A1B
                chainloader /efi/Microsoft/Boot/bootmgfw.efi
        }
      '';
    };
    efi.efiSysMountPoint = "/boot";
    efi.canTouchEfiVariables = true;
  };

  # Linux Kernel to use:
  boot.kernelPackages = pkgs.linuxPackages_latest;

  networking.hostName = "nixos"; # Define your hostname.
  #networking.wireless.enable = true;  # Enables wireless support via wpa_supplicant.
  networking.networkmanager.enable = true;
  networking.networkmanager.wifi.backend = "iwd";
  networking.nameservers = ["192.168.4.151"];
  networking.networkmanager.dns = "none";

  # Set your time zone.
  time.timeZone = "America/Los_Angeles";

  # The global useDHCP flag is deprecated, therefore explicitly set to false here.
  # Per-interface useDHCP will be mandatory in the future, so this generated config
  # replicates the default behaviour.
  networking.useDHCP = false;
  #networking.interfaces.enp38s0.useDHCP = true;
  #networking.interfaces.wlan0.useDHCP = true;

  # Configure network proxy if necessary
  # networking.proxy.default = "http://user:password@proxy:port/";
  # networking.proxy.noProxy = "127.0.0.1,localhost,internal.domain";

  # Select internationalisation properties.
  i18n.defaultLocale = "en_US.UTF-8";
  console = {
    font = "Lat2-Terminus16";
    keyMap = "us";
  };

  services.gnome.gnome-keyring.enable = true;
  programs.seahorse.enable = true;


  hardware.openrazer = {
   enable = true;
   users =[ "jrgiacone" ] ;
  };


  # Video Driver options: nvidia

   hardware = {
     opengl.enable = true;
     nvidia.nvidiaSettings = true;
     #nvidia.package = config.boot.kernelPackages.nvidiaPackages.beta;
     nvidia.powerManagement.enable = true;
     #nvidia.modesetting.enable = true;
   };

  services = {
    fstrim.enable = true;
    xserver = {
      enable = true;
      videoDrivers = [ "nvidia" ];
      displayManager = {
        #sddm.enable = true;
        startx.enable = true;
        defaultSession = "none+xmonad";
      };
      #desktopManager.plasma5.enable = true;
      windowManager = {
        xmonad.enable = true;
        xmonad.enableContribAndExtras = true;
      };
    };
  };

  # programs.sway = {
  #   enable = true;
  #   wrapperFeatures.gtk = true; # so that gtk works properly
  #   extraPackages = with pkgs; [
  #     swaylock
  #     swayidle
  #     wl-clipboard
  #     mako # notification daemon
  #     alacritty # Alacritty is the default terminal in the config
  #     dmenu # Dmenu is the default in the config but i recommend wofi since its wayland native
  #   ];
  # };

  # Configure keymap in X11
  services.xserver.layout = "us";
  # services.xserver.xkbOptions = "eurosign:e";

  # Enable CUPS to print documents.
  # services.printing.enable = true;

  # Enable sound.
  #sound.enable = true;
  #hardware.pulseaudio.enable = true;

  #hardware.bluetooth.enable = true;

  # rtkit is optional but recommended
  security.rtkit.enable = true;
  services.pipewire = {
    enable = true;
    alsa.enable = true;
    alsa.support32Bit = true;
    pulse.enable = true;
    # If you want to use JACK applications, uncomment this
    #jack.enable = true;
  };

  # Enable touchpad support (enabled default in most desktopManager).
  # services.xserver.libinput.enable = true;

  services.xserver.libinput.mouse.accelProfile = "flat";

  # Define a user account. Don't forget to set a password with ‘passwd’.
  users.users.jrgiacone = {
    isNormalUser = true;
    extraGroups = [ "wheel" "plugdev" "networkmanager" ]; # Enable ‘sudo’ for the user.
  };

  programs.dconf.enable = true;

  nixpkgs.config.packageOverrides = pkgs: {
    nur = import (builtins.fetchTarball "https://github.com/nix-community/NUR/archive/master.tar.gz") {
      inherit pkgs;
    };
  };

  # Steam options:
  programs.steam.enable = true;
  hardware.steam-hardware.enable = true;
  hardware.opengl.driSupport32Bit = true;
  programs.gamemode.enable = true;

  nixpkgs.config.allowUnfree = true;
  nix.extraOptions = '' experimental-features = nix-command '';

  nix.autoOptimiseStore = true;

  fonts.fonts = with pkgs; [
    font-awesome
  ];
  # List packages installed in system profile. To search, run:
  # $ nix search wget
  environment.systemPackages = with pkgs; [

    # GUI Apps:
    keepassxc
    firefox
    #bottles
    #heroic
    nur.repos.wolfangaukang.heroic
    nur.repos.wolfangaukang.gogdl
    #spotify
    chromium
    discord
    vscode
    pavucontrol
    easyeffects
    (pkgs.libsForQt5.callPackage /etc/nixos/polychromatic.nix { inherit (pkgs) meson wrapGAppsHook; })

    # TUI Apps:
    nvtop-nvidia
    ncspot
    htop
    vifm
    ncdu

    # Terminals:
    alacritty

    # Development:
    git
    gh
    #python37

    # Tools:
    udiskie
    playerctl
    neofetch
    neovim
    vim
    unzip
    wget
    efibootmgr
    xdotool
    vulkan-loader
    mangohud
    libsecret
    libgnome-keyring
    imagemagick

    # Window Manager Essentials:
    xmobar
    dmenu
    rofi
    trayer
    picom
    nitrogen
    dunst
    libnotify
    networkmanagerapplet
    hicolor-icon-theme

    # Gaming Test
    #wineWowPackages.staging
    wine
    winetricks
    wine-staging

  ];

  #services.flatpak.enable = true;
  #xdg.portal.enable = true;


  # Some programs need SUID wrappers, can be configured further or are
  # started in user sessions.
  # programs.mtr.enable = true;
  # programs.gnupg.agent = {
  #   enable = true;
  #   enableSSHSupport = true;
  # };

  # List services that you want to enable:

  # Enable the OpenSSH daemon.
  # services.openssh.enable = true;

  # Open ports in the firewall.
  # networking.firewall.allowedTCPPorts = [ ... ];
  # networking.firewall.allowedUDPPorts = [ ... ];
  # Or disable the firewall altogether.
  # networking.firewall.enable = false;

  # This value determines the NixOS release from which the default
  # settings for stateful data, like file locations and database versions
  # on your system were taken. It‘s perfectly fine and recommended to leave
  # this value at the release version of the first install of this system.
  # Before changing this value read the documentation for this option
  # (e.g. man configuration.nix or on https://nixos.org/nixos/options.html).
  system.stateVersion = "21.11"; # Did you read the comment?
```

**Comment by Icy-Thought** [comment]
@jrgiacone your setup shares a lot or packages in common with mine and therefore I don't think I have a "special" package installed which happens to enable polychromatic to function as expected.

But for the sake of testing, use taffybar (the bar I am currently using) and report back if it works or fails like it did with dmenu.

**Comment by piegamesde** [comment]
It looks like something went wrong while rebasing. Given that you can't un-ping people on GitHub, I'm going to close this now to prevent further spamming. Please open a new PR, and ping previous reviewers in there.

**Pull request closed without merging**